raidboss: Initial triggers and timeline for EW final trial

valarnin opened the pull request:
Triggers need more testing by someone other than me. There's probably issues with the `Doomed Stars` and `Epigonoi` triggers still.
Timeline is really basic and the loops for add phase and final phase aren't working, I'm probably doing something wrong there.

JLGarber commented on ui/raidboss/data/06-ew/trial/endsinger.ts:
```diff
@@ -0,0 +1,231 @@
+      id: 'Endsinger Death\'s Embrace Feathers',
+      type: 'StartsUsing',
+      netRegex: NetRegexes.startsUsing({ id: '6649', source: 'The Endsinger', capture: false }),
+      delaySeconds: 5.7,
```
Is there a reason to write the trigger against `startsUsing` here rather than just against `ability`? From the One Run's worth of log I have, it seems this one is always 5.7 seconds.

(Alternatively, what about `delayseconds: (_, matches) => parseFloat(matches.duration) - 0,`? This makes it immediately clear to the reader how soon before the mechanic resolution the trigger activates, rather than "here's a random-looking delay"

JLGarber commented on ui/raidboss/data/06-ew/trial/endsinger.ts:
```diff
@@ -0,0 +1,231 @@
+      id: 'Endsinger Hubris',
+      type: 'StartsUsing',
+      netRegex: NetRegexes.startsUsing({ id: '6652', source: 'The Endsinger', capture: true }),
+      response: Responses.tankBuster(),
```
My log has at least one instance of this being a `22` log line. Maybe it should be `tankCleave` instead?

JLGarber commented on ui/raidboss/data/06-ew/trial/endsinger.txt:
```diff
@@ -0,0 +1,107 @@
+1045.2 "Crash" sync / 1[56]:[^:]*:Kakodaimon:6657:/
+1050.2 "Crash" sync / 1[56]:[^:]*:Kakodaimon:6657:/
+1056.6 "Meteor Outburst" sync / 1[56]:[^:]*:Kakodaimon:665C:/
+1067.8 "Katastrophe" sync / 1[56]:[^:]*:Kakodaimon:665B:/ window 0,30
+1078.9 "Katastrophe" sync / 1[56]:[^:]*:Kakodaimon:665B:/
+1090.0 "Katastrophe" sync / 1[56]:[^:]*:Kakodaimon:665B:/
```
What I would normally do for something like this is to have 4-6 lines written out, and then jump from the second to the first. So it might look like this:

```
1067.8 "Katastrophe" sync / 1[56]:[^:]*:Kakodaimon:665B:/
1078.9 "Katastrophe" sync / 1[56]:[^:]*:Kakodaimon:665B:/ jump 1067.8
1090.0 "Katastrophe"
1101.1 "Katastrophe"
1112.2 "Katastrophe"
1123.3 "Katastrophe"
```

Technically we can probably get by with just four lines here, since 30s in the future is the max displayed normally, but I usually just make sure unreachable lookahead sections have at least six lines, including the jump line.

quisquous commented:
Thanks for this.  I'll take a closer look at this tomorrow.

JLGarber commented on ui/raidboss/data/06-ew/trial/endsinger.txt:
```diff
@@ -0,0 +1,107 @@
+2036.0 "--untargetable--"
+2074.0 "--targetable--"
+2086.1 "Telos" sync / 1[56]:[^:]*:The Endsinger:665E:/
+2099.3 "Telomania (big)" sync / 1[56]:[^:]*:The Endsinger:665F:/ window 0,30
+2099.9 "Telomania (small)" sync / 1[56]:[^:]*:The Endsinger:665F:/
```
Is this actually one big hit and one small hit from the same `665F` ability?

valarnin replied:
I got the order of hits wrong here, it's 4 small hits then the big hit, the first cast is purely a cast and no damage goes out.

JLGarber commented on ui/raidboss/data/06-ew/trial/endsinger.txt:
```diff
@@ -0,0 +1,107 @@
+2086.1 "Telos" sync / 1[56]:[^:]*:The Endsinger:665E:/
+2099.3 "Telomania (big)" sync / 1[56]:[^:]*:The Endsinger:665F:/ window 0,30
+2099.9 "Telomania (small)" sync / 1[56]:[^:]*:The Endsinger:665F:/
+2103.6 "Telomania (small)" sync / 1[56]:[^:]*:The Endsinger:6660:/
+2105.1 "Telomania (small)" sync / 1[56]:[^:]*:The Endsinger:6660:/
+2106.5 "Telomania (small)" sync / 1[56]:[^:]*:The Endsinger:6660:/
+2107.9 "Telomania (small)" sync / 1[56]:[^:]*:The Endsinger:6660:/
```
These are too close together to be synced. The default window is 2.5,2.5, so we will have collisions. What I normally do here is something like

```suggestion
2103.6 "Telomania (small) x4" duration 5 #sync / 1[56]:[^:]*:The Endsinger:6660:/
```

This gives healers a feel for how long they have to stay focused on small hits, and we avoid weird timeline jumps from multiple simultaneous syncs. (We comment the sync rather than removing it entirely, since this entry is reachable, it just shouldn't sync.)

JLGarber commented on ui/raidboss/data/06-ew/trial/endsinger.txt:
```diff
@@ -0,0 +1,107 @@
+2105.1 "Telomania (small)" sync / 1[56]:[^:]*:The Endsinger:6660:/
+2106.5 "Telomania (small)" sync / 1[56]:[^:]*:The Endsinger:6660:/
+2107.9 "Telomania (small)" sync / 1[56]:[^:]*:The Endsinger:6660:/
+2114.6 "Telomania (small)" sync / 1[56]:[^:]*:The Endsinger:6661:/
+2120.1 "Dead Star" sync / 1[56]:[^:]*:The Endsinger:5E4E:/
+2122.1 "Dead Star" sync / 1[56]:[^:]*:The Endsinger:5E4E:/
+2124.1 "Dead Star" sync / 1[56]:[^:]*:The Endsinger:5E4E:/
+2124.5 "Telomania (big)" sync / 1[56]:[^:]*:The Endsinger:665F:/
+2125.1 "Telomania (small)" sync / 1[56]:[^:]*:The Endsinger:665F:/
```
The big hit at 2124.5 is probably where you want to put a `jump 2099.3`. That will give you a clean loop. You could then remove the syncs from the remaining entries to indicate that the remainder is for timeline visuals only and should never be reached.

(Additionally, you might want to build out this second fake loop by incrementing the first loop by 25.2. This will ensure the fake loop exactly matches with the main loop, and there will be no tiny shudders when the jump occurs.)

JLGarber reviewed (commented):
Thanks so much for putting this together, valarnin. This all looked daunting to poke through, so I'm glad that's been done. I certainly wasn't looking forward to having to run it multiple times with randoms!

valarnin commented:
Thanks for the feedback from JLGarber.

From running through again today after adjustments, I have the following remaining issues:
- Timeline doesn't sync properly for the first Elegeia but seems to figure itself out eventually, no clue why. I'm not very good with timelines yet.
- Doomed Stars seems OK for the 3 variations I've gotten so far.
- The replay mechanic is not consistent for the two-orb replay, so I think that logic is going to need to be changed (`Endsinger Fatalism`) to look for the tether instead?

quisquous commented on ui/raidboss/data/06-ew/trial/endsinger.ts:
```diff
@@ -0,0 +1,251 @@
+
+        if (x < 100) {
+          if (y < 100)
+            return output.se?.();
```
Can you use `output.se!()` here? When I converted all the triggers to TypeScript, there was some discussion and we decided that output functions were the one place to use non-null assertions just to make them slightly less obnoxious to read.  If in the future TypeScript can understand more about proxies then we won't need it.

quisquous commented on ui/raidboss/data/06-ew/trial/endsinger.ts:
```diff
@@ -0,0 +1,251 @@
+    {
+      id: 'Endsinger Ultimate Fate',
+      type: 'GameLog',
+      netRegex: NetRegexes.gameLog({ code: '0039', line: '4 seconds until oblivion is sealed\\.\\.\\.' }),
```
What about making this Ultimate Fate (6B59) starts casting? In general I would prefer to avoid game logs as much as possible.  It's extra translation work, and there's no end of people who don't have game logs for various reasons.

```
00|2021-12-08T18:04:27.0000000-08:00|0039||4 seconds until oblivion is sealed...|bab2a4e54fc622fb
20|2021-12-08T18:04:27.9420000-08:00|4000F88A|The Endsinger|6B59|Ultimate Fate|4000F88A|The Endsinger|4.70|100.00|100.00|0.00|-3.14|92bb7e981862f743
```

(Also, CI is complaining about missing `capture: false` on this line)

valarnin replied:
I tried this with startsCasting and was having issues with it. I'll revisit it tomorrow, probably was a PEBKAC error.

quisquous commented on ui/raidboss/data/06-ew/trial/endsinger.txt:
```diff
@@ -0,0 +1,100 @@
+
+# P1
+0.0 "--sync--" sync /Engage!/ window 0,1
+13.7 "Elegeia" sync / 1[56]:[^:]*:The Endsinger:662C:/ window 15,15
```
```suggestion
8.7 "--sync--" sync / 14:[^:]*:The Endsinger:662C:/ window 15,15
13.7 "Elegeia" sync / 1[56]:[^:]*:The Endsinger:662C:/
```

If there's no zone closing, it's nice to sync to the starts casting so the timeline starts sooner.

quisquous commented on ui/raidboss/data/06-ew/trial/endsinger.txt:
```diff
@@ -0,0 +1,100 @@
+1218.9 "--targetable--"
+
+# Phase 3, loops until soft enrage
+2000.0 "Fatalism" sync / 1[56]:[^:]*:The Endsinger:6B5B:/ window 1000,0
+2036.0 "--untargetable--"
+2074.0 "--targetable--"
+2086.1 "Telos" sync / 1[56]:[^:]*:The Endsinger:665E:/
+2099.3 "Telomania (cast)" sync / 1[56]:[^:]*:The Endsinger:665F:/ window 0,30
+2103.6 "Telomania (small) x4" duration 5 #sync / 1[56]:[^:]*:The Endsinger:6660:/
+2114.6 "Telomania (big)" sync / 1[56]:[^:]*:The Endsinger:6661:/
+2120.1 "Dead Star x3" duration 6 #sync / 1[56]:[^:]*:The Endsinger:5E4E:/
+2124.5 "Telomania (cast)" sync / 1[56]:[^:]*:The Endsinger:665F:/ window 0,30 jump 2099.3
+2128.8 "Telomania (small) x4" duration 5 #sync / 1[56]:[^:]*:The Endsinger:6660:/
+2139.8 "Telomania (big)" #sync / 1[56]:[^:]*:The Endsinger:6661:/
+2145.3 "Dead Star x3" duration 6 #sync / 1[56]:[^:]*:The Endsinger:5E4E:/
```
Can you add a loop block underneath dead star, rather than a huge sync on Fatalism? Otherwise, the timeline will be blank and then suddenly fill in again.  It's also not clear how much time there is between Dead Star and the next use of Fatalism.  (Maiko's extension in VSCode has a "set time to", where you can copy this whole block, select the copy, and then right click and say "set time to", and it will set the first time and adjust everything after it to easily make a loop block.)

e.g. https://github.com/quisquous/cactbot/blob/main/ui/raidboss/data/06-ew/dungeon/the_tower_of_zot.txt#L74-L79

JLGarber replied:
I think this *is* the loop block, it's just not delineated with whitespace. There should probably be a blank line between 2086.1/2099.3, and between 2120.1/2124.5. The Fatalism sync is on the other side of the block.

quisquous replied:
Oh, you're right, thanks!

quisquous commented on ui/raidboss/data/06-ew/trial/endsinger.ts:
```diff
@@ -0,0 +1,251 @@
+      id: 'Endsinger Ultimate Fate',
+      type: 'GameLog',
+      netRegex: NetRegexes.gameLog({ code: '0039', line: '4 seconds until oblivion is sealed\\.\\.\\.' }),
+      suppressSeconds: 65,
```
Why is there a suppress?

valarnin replied:
The line is repeated for the recast of the ability, which can't be tank LB'd, so there's no point in having the trigger fire again.

quisquous commented on ui/raidboss/data/06-ew/trial/endsinger.ts:
```diff
@@ -0,0 +1,251 @@
+      id: 'Endsinger Elenchos Outsides',
+      type: 'StartsUsing',
+      netRegex: NetRegexes.startsUsing({ id: '6642', source: 'The Endsinger', capture: false }),
+      response: Responses.getIn(),
```
This is my own personal preferences, but "middle" vs "sides" is for stuff like this, and "in" vs "out" is for donut/circle mechanics.

```suggestion
      response: Responses.getMiddle(),
```

valarnin replied:
I think I started with sides, and tried `getSides` and it didn't exist, and didn't think to check for `goSides`.

quisquous commented on ui/raidboss/data/06-ew/trial/endsinger.ts:
```diff
@@ -0,0 +1,251 @@
+      id: 'Endsinger Elenchos Middle',
+      type: 'StartsUsing',
+      netRegex: NetRegexes.startsUsing({ id: '6644', source: 'The Endsinger', capture: false }),
+      response: Responses.getOut(),
```
```suggestion
      response: Responses.goSides(),
```

quisquous commented on ui/raidboss/data/06-ew/trial/endsinger.ts:
```diff
@@ -0,0 +1,251 @@
+      },
+      outputStrings: {
+        awayFromHead: {
+          en: 'Away from head',
```
Maybe `Avoid Head Lasers`? To imply that it's a line aoe and not a distance thing?

quisquous commented on ui/raidboss/data/06-ew/trial/endsinger.txt:
```diff
@@ -0,0 +1,100 @@
+# -oc 'The Endsinger' 'Kakodaimon'
+# -ii 6281 6C61 6660 6661 6C6A 664B 6653 6643 67FA 6642 6648 67FB 6655
+
+# Not sure if/where P1 loops
```
I have this:
```
490.4 "--untargetable--"
499.7 "Interstellar" sync / 1[56]:[^:]*:The Endsinger:664D:/
501.8 "Nemesis" sync / 1[56]:[^:]*:The Endsinger:664E:/
507.8 "--targetable--"
514.0 "--sync--" sync / 1[56]:[^:]*:The Endsinger:664A:/
520.0 "Death's Embrace" sync / 1[56]:[^:]*:The Endsinger:6649:/
528.8 "Pharmakon" sync / 1[56]:[^:]*:Feather of Despair:664C:/
539.2 "Hubris" sync / 1[56]:[^:]*:The Endsinger:6652:/
549.4 "Aporrhoia" sync / 1[56]:[^:]*:The Endsinger:663D:/
559.0 "Elenchos" sync / 1[56]:[^:]*:The Endsinger:663E:/
576.8 "Elegeia" sync / 1[56]:[^:]*:The Endsinger:662C:/
583.2 "--sync--" sync / 1[56]:[^:]*:Doomed Stars:662D:/
590.2 "Stellar Collision" sync / 1[56]:[^:]*:Doomed Stars:662E:/
596.0 "Nemesis" sync / 1[56]:[^:]*:The Endsinger:6654:/
603.1 "--untargetable--"
612.4 "Interstellar" sync / 1[56]:[^:]*:The Endsinger:664D:/
614.6 "Nemesis" sync / 1[56]:[^:]*:The Endsinger:664E:/
620.6 "--targetable--"
626.7 "--sync--" sync / 1[56]:[^:]*:The Endsinger:664A:/
632.7 "Death's Embrace" sync / 1[56]:[^:]*:The Endsinger:6649:/
641.5 "Pharmakon" sync / 1[56]:[^:]*:Feather of Despair:664C:/
642.8 "Elenchos" sync / 1[56]:[^:]*:The Endsinger:6644:/
651.9 "Hubris" sync / 1[56]:[^:]*:The Endsinger:6652:/
662.1 "Aporrhoia" sync / 1[56]:[^:]*:The Endsinger:663D:/
671.7 "Elenchos" sync / 1[56]:[^:]*:The Endsinger:663E:/
672.8 "Liturgy of the Bell" sync / 1[56]:[^:]*:Liturgic Bell:6508:/
689.6 "Elegeia" sync / 1[56]:[^:]*:The Endsinger:662C:/
696.0 "--sync--" sync / 1[56]:[^:]*:Doomed Stars:662D:/
703.0 "Stellar Collision" sync / 1[56]:[^:]*:Doomed Stars:662E:/
708.8 "Nemesis" sync / 1[56]:[^:]*:The Endsinger:6654:/
715.9 "--untargetable--"
725.2 "Interstellar" sync / 1[56]:[^:]*:The Endsinger:664D:/
727.3 "Nemesis" sync / 1[56]:[^:]*:The Endsinger:664E:/
733.3 "--targetable--"
739.5 "--sync--" sync / 1[56]:[^:]*:The Endsinger:664A:/
745.5 "Death's Embrace" sync / 1[56]:[^:]*:The Endsinger:6649:/
754.3 "Pharmakon" sync / 1[56]:[^:]*:Feather of Despair:664C:/
755.6 "Elenchos" sync / 1[56]:[^:]*:The Endsinger:6644:/
```

It seems possible that 745.5 loops back to 632.7?

JLGarber replied:
I think you're right. It would be highly unusual for a normal-mode trial to have a phase 1 loop, but definitely not impossible.

quisquous replied:
Oh no, I take this back, I'm not sure what's going on here.  I've added in some more lines above, but when I run it through test_timeline, I get this:

```
    Missed sync: Aporrhoia at 436.6 (last seen at 285.68)
    Missed sync: Elenchos at 450.8 (last seen at 417.3)
    Missed sync: Elenchos at 563.5 (last seen at 417.3)
```

In particular, I don't see the double Elenchos at 450.8 or 563.5, and I'm missing that Aporrhoia entirely? I don't see it in make_timeline either on my log.  (Happy to share it if it's more helpful than just the above make_timeline output.)

valarnin replied:
I'm so sorry you had a P1 last that long... The longest I've had was only about 600 seconds.

It might be a two-pattern loop, e.g. death => hubris is one pattern, and death => elenchos is the other? This is pretty beyond my knowledge of timelines to figure out/set up though.

quisquous replied:
> I'm so sorry you had a P1 last that long...

Me tooooo

> It might be a two-pattern loop, e.g. death => hubris is one pattern, and death => elenchos is the other?

Yeah, this is plausible.  Feel free to leave this as a TODO for now and with more data we can fix it.  When I left the initial comment I thought maybe this could be fixed up more easily, but now it seems more complicated.

quisquous commented on ui/raidboss/data/06-ew/trial/endsinger.txt:
```diff
@@ -0,0 +1,100 @@
+2103.6 "Telomania (small) x4" duration 5 #sync / 1[56]:[^:]*:The Endsinger:6660:/
+2114.6 "Telomania (big)" sync / 1[56]:[^:]*:The Endsinger:6661:/
+2120.1 "Dead Star x3" duration 6 #sync / 1[56]:[^:]*:The Endsinger:5E4E:/
+2124.5 "Telomania (cast)" sync / 1[56]:[^:]*:The Endsinger:665F:/ window 0,30 jump 2099.3
```
Can you add a tiny bit of time prior to this? On the off chance that the timeline is slightly behind, you don't want this jump to be missed.

```suggestion

2124.5 "Telomania (cast)" sync / 1[56]:[^:]*:The Endsinger:665F:/ window 10,30 jump 2099.3
```

quisquous commented on ui/raidboss/data/06-ew/trial/endsinger.txt:
```diff
@@ -0,0 +1,100 @@
+2036.0 "--untargetable--"
+2074.0 "--targetable--"
+2086.1 "Telos" sync / 1[56]:[^:]*:The Endsinger:665E:/
+2099.3 "Telomania (cast)" sync / 1[56]:[^:]*:The Endsinger:665F:/ window 0,30
```
```suggestion
2099.3 "Telomania (cast)" sync / 1[56]:[^:]*:The Endsinger:665F:/
```

I think no need for a window here, as well, since you have the jump below.

quisquous commented:
Also, thank you so much for this! I really appreciate all of this and am glad somebody else figured out what to do with the planets.  :)

valarnin commented:
Alright, rebased to main and fixed a ton of stuff. I think this is ready for final review and merge now. The only thing that might need tweaked is the `Interstellar` trigger's `delaySeconds` timing.

quisquous commented:
Looks great, thanks! I think if the interstellar timing could be better, that's something that could be tweaked later.  How minimal of a delay can there be there?

valarnin commented:
> How minimal of a delay can there be there?

I didn't test it for very long, just found a value that worked. This is one of those edge cases where raidemulator can't directly help because there's no log line for the boss here to signify when they've fully teleported over.
